File: src/test/java/pages/geographypage/GeographyMainPage.java
package pages.geographypage;

import com.microsoft.playwright.Locator;
import com.microsoft.playwright.Page;
import com.microsoft.playwright.TimeoutError;
import com.microsoft.playwright.options.WaitForSelectorState;
import pages.base.BasePage;

/**
 * Geography Main Page - Profile Selection and Geography Set Search
 * Screen: Geographies Main View
 */
public class GeographyMainPage extends BasePage {

    // Locators - Profile Selection
    private static final String PROFILE_LIST = "//*[@id='profileSelect_list']/p-selectitem/li";
    private static final String SEARCH_PROFILE = "//*[@id='profileSelect']";
    private static final String INPUT_SEARCH_PROFILE = "//*[@id='profileSelect']//input";
    private static final String CREATE_NEW_GEO_BTN = " Create New Geography Set ";

    // Locators - Geography Set
    private static final String GEOGRAPHY_SET_LIST = "//div[@class='col-md-6 col-xl-4 mb-4 ng-star-inserted']";
    private static final String SEARCH_GEO_SETS_PLACEHOLDER = "Search geography sets...";

    // Locators - Region Selection
    private static final String CUSTOM_REGION = "//h6[text()='Custom Region']";
    private static final String CUSTOM_CENSUS_CHAIN = "//h6[text()='Custom Census Chain (CCC)']";

    public GeographyMainPage(Page page) {
        super(page);
    }

    // ============= Navigation Validation =============

    public boolean isOnGeographyPage() {
        page.waitForURL("**/geographies**");
        return page.url().contains("geographies");
    }

    // ============= Profile Management =============

    public int getProfileCount() {
        return getElementCount(PROFILE_LIST);
    }

    public void openProfileSearch() {
        clickElement(SEARCH_PROFILE);
    }

    public void searchProfile(String profileName) {
        if (profileName == null || profileName.trim().isEmpty()) {
            throw new IllegalArgumentException("Profile name cannot be empty");
        }
        clickAndEnter(INPUT_SEARCH_PROFILE, profileName);
    }

    public void selectProfile(int index) {
        try {
            page.waitForSelector(PROFILE_LIST);
            page.locator(PROFILE_LIST).nth(index).click();
        } catch (TimeoutError e) {
            throw new RuntimeException("Failed to select profile at index: " + index);
        }
    }

    public boolean isProfileSelected() {
        return isElementVisible(CREATE_NEW_GEO_BTN);
    }

    // ============= Geography Set Search & Selection =============

    public int getGeographySetsCount() {
        waitForStabilization();
        return getElementCount(GEOGRAPHY_SET_LIST);
    }

    public void searchGeographySet(String geoSetName) {
        if (geoSetName == null || geoSetName.trim().isEmpty()) {
            throw new IllegalArgumentException("Geography set name cannot be empty");
        }
        fillByPlaceholder(SEARCH_GEO_SETS_PLACEHOLDER, geoSetName);
    }

    public void selectGeographySetByName(String setName) {
        String selector = buildCardSelector(setName, "h5");
        try {
            page.waitForSelector("div.geography-set-card",
                new Page.WaitForSelectorOptions()
                    .setState(WaitForSelectorState.VISIBLE)
                    .setTimeout(DEFAULT_TIMEOUT));
            
            Locator card = page.locator(selector);
            if (card.count() == 0) {
                throw new RuntimeException("Geography set not found: " + setName);
            }
            card.first().click();
        } catch (TimeoutError e) {
            throw new RuntimeException("Timeout waiting for geography set: " + setName);
        }
    }

    public void selectGeographySetByStatus(String status) {
        String selector = buildCardSelector(status, "span");
        try {
            page.waitForSelector("div.geography-set-card",
                new Page.WaitForSelectorOptions()
                    .setState(WaitForSelectorState.VISIBLE)
                    .setTimeout(DEFAULT_TIMEOUT));
            
            Locator card = page.locator(selector);
            if (card.count() == 0) {
                throw new RuntimeException("Geography set with status not found: " + status);
            }
            card.first().click();
        } catch (TimeoutError e) {
            throw new RuntimeException("Timeout waiting for geography set with status: " + status);
        }
    }

    private String buildCardSelector(String text, String tagName) {
        return "//div[contains(@class,'geography-set-card')]//" + tagName +
               "[translate(normalize-space(text()), 'ABCDEFGHIJKLMNOPQRSTUVWXYZ', 'abcdefghijklmnopqrstuvwxyz') = '" +
               text.toLowerCase() + "']/ancestor::div[contains(@class,'geography-set-card')]";
    }

    // ============= Geography Set Creation =============

    public void clickCreateNewGeographySet() {
        clickElementByText(CREATE_NEW_GEO_BTN);
        handleGeographyTypeSelection();
    }

    private void handleGeographyTypeSelection() {
        if (isElementVisible(CUSTOM_REGION) && isElementVisible(CUSTOM_CENSUS_CHAIN)) {
            clickElement(CUSTOM_CENSUS_CHAIN);
        }
    }
}

File: pages/geographyMainPage.ts
import { expect, Locator, Page } from '@playwright/test';
import { BasePage } from './basePage';

export class GeographyMainPage extends BasePage {
  readonly geographyNav: Locator;
  readonly searchProfileButton: Locator;
  readonly profileSearchInput: Locator;
  readonly profileRows: Locator;
  readonly geographySearchInput: Locator;
  readonly createNewGeographyButton: Locator;

  constructor(page: Page) {
    super(page);
    this.geographyNav = page.getByText(/Geography/i).first();
    this.searchProfileButton = page.locator('//*[@id="profileSelect"]');
    this.profileSearchInput = page.locator('//*[@id="profileSelect"]//input');
    this.profileRows = page.locator('//*[@id="profileSelect_list"]/p-selectitem/li');
    this.geographySearchInput = page.locator('//input[@placeholder="Search geography sets..."]');
    this.createNewGeographyButton = page.getByText('Create New Geography Set');
  }

  async openGeography(): Promise<void> {
    await this.geographyNav.click();
  }

  async isOnGeographyPage(): Promise<boolean> {
    await this.page.waitForURL('**/geographies**');
    return this.isOnScreenByURL('geographies');
  }

  async openProfileSearch(): Promise<void> {
    await this.searchProfileButton.click();
  }

  async searchProfile(profileName: string): Promise<void> {
    const input = this.profileSearchInput;
    await expect(input).toBeVisible({ timeout: 15000 });
    await input.fill(profileName);
    await input.press('Enter');
    await this.page.waitForTimeout(300);
  }

  async getProfileCount(): Promise<number> {
    return this.getElementCount('//*[@id="profileSelect_list"]/p-selectitem/li');
  }

  async selectProfileByName(profileName: string): Promise<void> {
    const profileOption = this.page.locator('//*[@id="profileSelect_list"]/p-selectitem/li').filter({ hasText: profileName });
    await expect(profileOption.first()).toBeVisible({ timeout: 15000 });
    await profileOption.first().click();
  }

  async selectFirstProfile(): Promise<void> {
    const firstProfile = this.page.locator('//*[@id="profileSelect_list"]/p-selectitem/li').first();
    // const firstProfile = this.page.locator('//*[@id="profileSelect_list"]/p-selectitem/li').nth(2);
    await expect(firstProfile).toBeVisible({ timeout: 15000 });
    await firstProfile.click();
  }

  async selectProfile(index: number): Promise<void> {
    await this.page.locator('//*[@id="profileSelect_list"]/p-selectitem/li').nth(index).click();
  }

  async searchGeographySet(term: string): Promise<void> {
    await this.geographySearchInput.waitFor({ state: 'visible', timeout: 10000 });
    await this.geographySearchInput.fill(term);
  }

  async selectGeographySetByName(name: string): Promise<void> {
    await this.page.getByText(name).first().click();
  }

  async clickCreateNewGeographySet(): Promise<void> {
    await this.createNewGeographyButton.click();
  }
}
